### ISSUE TITLE
Geth Crash - program exceeds 10000-thread limit

### ISSUE THREAD

**Issue by rfikki** [ethereum/go-ethereum#2212]
```text
runtime: program exceeds 10000-thread limit
fatal error: thread exhaustion

runtime stack:
runtime.throw(0xf07990, 0x11)
	c:/go/src/runtime/panic.go:527 +0x97
runtime.checkmcount()
	c:/go/src/runtime/proc1.go:93 +0x9b
runtime.mcommoninit(0xc0ab9df680)
	c:/go/src/runtime/proc1.go:113 +0x100
runtime.allocm(0xc082012000, 0x10805e8, 0x449d6d)
	c:/go/src/runtime/proc1.go:874 +0xaa
runtime.newm(0x10805e8, 0xc082012000)
	c:/go/src/runtime/proc1.go:1091 +0x36
runtime.startm(0xc082012000, 0x100000001)
	c:/go/src/runtime/proc1.go:1178 +0xf6
runtime.handoffp(0xc082012000)
	c:/go/src/runtime/proc1.go:1207 +0xdf
runtime.retake(0x51727d1bb55c, 0x0)
	c:/go/src/runtime/proc1.go:3104 +0x1e7
runtime.sysmon()
	c:/go/src/runtime/proc1.go:3031 +0x216
runtime.mstart1()
	c:/go/src/runtime/proc1.go:721 +0xf1
runtime.mstart()
	c:/go/src/runtime/proc1.go:691 +0x79

...
```

**Comment by fjl** [comment]
Which version of geth is crashing? Does it happen repeatedly or just once?

**Comment by rfikki** [comment]
Geth/v1.4.0-unstable/windows/go1.5.1

This only happened once.

**Comment by fjl** [comment]
One more question: is this the entire backtrace dump or just the first few lines?
I think it would help to have all of it for this issue.

**Comment by rfikki** [comment]
File is extremely large over 30MB

**Comment by fjl** [comment]
Can you try to compress it, would that work?

**Comment by rfikki** [comment]
[01_Frontier_Morden_Console_crash.zip](https://github.com/ethereum/go-ethereum/files/129852/01_Frontier_Morden_Console_crash.zip)

**Comment by fjl** [comment]
Thanks. I have found the issue, will fix it for 1.4.

The issue is that there are 10000 goroutines doing this:

```text
goroutine 112085 [runnable, locked to thread]:
syscall.Syscall9(0x7fffba58d8b0, 0x7, 0xc0ab8cbd80, 0x80000000, 0x3, 0x0, 0x3, 0x80, 0x0, 0x0, ...)
	c:/go/src/runtime/syscall_windows.go:150 +0x5c
syscall.CreateFile(0xc0ab8cbd80, 0x380000000, 0x0, 0x8000000003, 0x0, 0x0, 0x0, 0x0)
	c:/go/src/syscall/zsyscall_windows.go:266 +0xc2
syscall.Open(0xc082230cc0, 0x38, 0x80000, 0x0, 0x7, 0x0, 0x0)
	c:/go/src/syscall/syscall_windows.go:293 +0x1b9
os.openFile(0xc082230cc0, 0x38, 0x0, 0x0, 0x107aa6e, 0x0, 0x0)
	c:/go/src/os/file_windows.go:85 +0xa1
os.OpenFile(0xc082230cc0, 0x38, 0x0, 0xc000000000, 0xc000000000, 0x0, 0x0)
	c:/go/src/os/file_windows.go:141 +0x1db
os.Open(0xc082230cc0, 0x38, 0xc000000000, 0x0, 0x0)
	c:/go/src/os/file.go:246 +0x4f
io/ioutil.ReadDir(0xc082230cc0, 0x38, 0x0, 0x0, 0x0, 0x0, 0x0)
	c:/go/src/io/ioutil/ioutil.go:101 +0x6b
github.com/ethereum/go-ethereum/crypto.getKeyAddresses(0xc082230cc0, 0x38, 0x0, 0x0, 0x0, 0x0, 0x0)
	D:/builds/geth_win32/build-go-win-develop/go/src/github.com/ethereum/go-ethereum/crypto/key_store_plain.go:196 +0x77
github.com/ethereum/go-ethereum/crypto.keyStorePassphrase.GetKeyAddresses(0xc082230cc0, 0x38, 0x40000, 0x1, 0x8, 0x20, 0x0, 0x0, 0x0, 0x0, ...)
	D:/builds/geth_win32/build-go-win-develop/go/src/github.com/ethereum/go-ethereum/crypto/key_store_passphrase.go:94 +0x62
github.com/ethereum/go-ethereum/crypto.(*keyStorePassphrase).GetKeyAddresses(0xc082224db0, 0x0, 0x0, 0x0, 0x0, 0x0)
	<autogenerated>:4 +0xc4
github.com/ethereum/go-ethereum/accounts.(*Manager).Accounts(0xc082224e10, 0x0, 0x0, 0x0, 0x0, 0x0)
	D:/builds/geth_win32/build-go-win-develop/go/src/github.com/ethereum/go-ethereum/accounts/account_manager.go:185 +0x85
github.com/ethereum/go-ethereum/eth.(*PublicBlockChainAPI).doCall(0xc0996f83c0, 0x0, 0x0, 0x2fedecf400000000, 0x892d6ff933e32c68, 0xdd95fcd8de13c666, 0x0, 0x0, 0x0, 0x0, ...)
	D:/builds/geth_win32/build-go-win-develop/go/src/github.com/ethereum/go-ethereum/eth/api.go:508 +0x213
github.com/ethereum/go-ethereum/eth.(*PublicBlockChainAPI).Call(0xc0996f83c0, 0x0, 0x0, 0x2fedecf400000000, 0x892d6ff933e32c68, 0xdd95fcd8de13c666, 0x0, 0x0, 0x0, 0x0, ...)
	D:/builds/geth_win32/build-go-win-develop/go/src/github.com/ethereum/go-ethereum/eth/api.go:553 +0x7d
reflect.Value.call(0xd020c0, 0xe19990, 0x13, 0xe57d60, 0x4, 0xc0b0e1d400, 0x3, 0x3, 0x0, 0x0, ...)
	c:/go/src/reflect/value.go:432 +0x1211
reflect.Value.Call(0xd020c0, 0xe19990, 0x13, 0xc0b0e1d400, 0x3, 0x3, 0x0, 0x0, 0x0)
	c:/go/src/reflect/value.go:300 +0xb8
github.com/ethereum/go-ethereum/rpc.(*Server).handle(0xc0996f8240, 0x18e8720, 0xc0b08f2040, 0x18e8648, 0xc0b33fa230, 0xc097c927e0, 0x0, 0x0)
	D:/builds/geth_win32/build-go-win-develop/go/src/github.com/ethereum/go-ethereum/rpc/server.go:322 +0xf83
github.com/ethereum/go-ethereum/rpc.(*Server).exec(0xc0996f8240, 0x18e8720, 0xc0b08f2040, 0x18e8648, 0xc0b33fa230, 0xc097c927e0)
	D:/builds/geth_win32/build-go-win-develop/go/src/github.com/ethereum/go-ethereum/rpc/server.go:345 +0x243
created by github.com/ethereum/go-ethereum/rpc.(*Server).ServeCodec
	D:/builds/geth_win32/build-go-win-develop/go/src/github.com/ethereum/go-ethereum/rpc/server.go:175 +0x90a
```

The solution will be to cache account addresses properly instead of reading the directory all the time.
This will also fix #2100.

**Comment by rfikki** [comment]
Thanks for your quick response and possible fix.
